zkevm-circuits/src/bytecode_circuit/bytecode_unroller.rs

```diff
@@ -772,12 +772,103 @@ impl<F: Field> SubCircuit<F> for BytecodeCircuit<F> {
+    use log::error;
+
+    impl<F: Field> Circuit<F> for BytecodeCircuit<F> {
+        type Config = (BytecodeCircuitConfig<F>, Challenges);
+        type FloorPlanner = SimpleFloorPlanner;
+
+        fn without_witnesses(&self) -> Self {
+            Self::default()
+        }
+
+        fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+            let bytecode_table = BytecodeTable::construct(meta);
+            let keccak_table = KeccakTable::construct(meta);
+            let challenges = Challenges::construct(meta);
+
+            let config = {
+                let challenges = challenges.exprs(meta);
+                BytecodeCircuitConfig::new(
+                    meta,
+                    BytecodeCircuitConfigArgs {
+                        bytecode_table,
+                        keccak_table,
+                        challenges,
+                    },
+                )
+            };
+
+            (config, challenges)
+        }
+
+        fn synthesize(
+            &self,
+            (config, challenges): Self::Config,
+            mut layouter: impl Layouter<F>,
+        ) -> Result<(), Error> {
+            let challenges = challenges.values(&mut layouter);
+
+            config.keccak_table.dev_load(
+                &mut layouter,
+                self.bytecodes.iter().map(|b| &b.bytes),
+                &challenges,
+            )?;
+            self.synthesize_sub(&config, &challenges, &mut layouter)?;
+            Ok(())
+        }
+    }
+
+    impl<F: Field> BytecodeCircuit<F> {
+        /// Verify that the selected bytecode fulfills the circuit
+        pub fn verify_raw(k: u32, bytecodes: Vec<Vec<u8>>) {
+            let unrolled: Vec<_> = bytecodes.iter().map(|b| unroll(b.clone())).collect();
+            Self::verify(k, unrolled, true);
+        }
+
+        pub(crate) fn verify(k: u32, bytecodes: Vec<UnrolledBytecode<F>>, success: bool) {
+            let circuit = BytecodeCircuit::<F>::new(bytecodes, 2usize.pow(k));
+
+            let prover = MockProver::<F>::run(k, &circuit, Vec::new()).unwrap();
+            let result = prover.verify();
+            if let Err(failures) = &result {
+                for failure in failures.iter() {
+                    error!("{}", failure);
+                }
+            }
+            assert_eq!(result.is_ok(), success);
+        }
+    }
+
+    /// Test bytecode circuit with unrolled bytecode
+    pub fn test_bytecode_circuit_unrolled<F: Field>(
+        k: u32,
+        bytecodes: Vec<UnrolledBytecode<F>>,
+        success: bool,
+    ) {
+        let circuit = BytecodeCircuit::<F>::new(bytecodes, 2usize.pow(k));
+
+        let prover = MockProver::<F>::run(k, &circuit, Vec::new()).unwrap();
+        let result = prover.verify_par();
+        if let Err(failures) = &result {
+            for failure in failures.iter() {
+                error!("{}", failure);
+            }
+        }
+        assert_eq!(result.is_ok(), success);
+    }
```

Comment:
Why isn't all that stuff moved to a test_util.rs file or whatever so that we reduce the line ammount and split a bit more the code?

Reply:
Make sense.

Reply:
I've separated test-mod into per file.